stable/redis-ha/README.md

```diff
@@ -41,46 +41,32 @@ $ helm delete <chart-name>
-| `replicas.sentinels`             | Number of sentinel pods                                                                                                      | 3                                                         |
+| `image`                          | Redis image                                                                                                                  | `redis`                                                   |
+| `tag`                            | Redis tag                                                                                                                    | `4.0.11-stretch`                                          |
+| `redis.resources`                | CPU/Memory for master/slave nodes resource requests/limits                                                                   | Memory: `200Mi`, CPU: `100m`                              |
+| `sentinel.resources`             | CPU/Memory for sentinel node resource requests/limits                                                                        | Memory: `200Mi`, CPU: `100m`                              |
+| `replicas`                       | Number of redis master/slave pods                                                                                            | 3                                                         |
 | `nodeSelector`                   | Node labels for pod assignment                                                                                               | {}                                                        |
 | `tolerations`                    | Toleration labels for pod assignment                                                                                         | []                                                        |
-| `podAntiAffinity.sentinel`       | Antiaffinity for pod assignment of sentinels, `hard` or `soft`                                                               | `soft`                                                    |
 | `podAntiAffinity.server`         | Antiaffinity for pod assignment of servers, `hard` or `soft`                                                                 | `soft`                                                    |
-| `servers.serviceType`            | Set to "LoadBalancer" to enable access from the VPC                                                                          | ClusterIP                                                 |
-| `servers.annotations`            | See Appliance mode                                                                                                           | ``                                                        |
-| `rbac.create`                    |  whether RBAC resources should be created                                                                                    | true                                                      |
```

Comment:
Hi @ssalaues  it looks like this PR completely reverses https://github.com/helm/charts/pull/3405. Is there a reason for that?

Reply:
The previous custom image [smile:redis](https://github.com/smileisak/docker-images/tree/master/redis/alpine) had a requirement to write labels to the pods to denote their role (master, slave, sentinel) using [this script](https://github.com/smileisak/docker-images/blob/master/redis/alpine/label-updater.sh).

This implementation has no such requirement, but rather determines the role during pod init, and hands off leader election to sentinel. I am a little curious whether there's an opportunity to get into a split-brain scenario during pod churn if both the init container and sentinel are making leader election decisions.

Reply:
Thanks for the review @scottrigby !

So from my understanding of the redis-ha chart in its current state, it uses docker images built from the repo here https://github.com/smileisak/docker-images/tree/master/redis/alpine which uses an assortment of scripts to update labels, find master/slaves, and start elections/promotions via kubectl commands, which require the appropriate roles and accounts to function in a typical RBAC enabled environment. Which is super cool in theory but has been the source of many issues (multiple masters being labeled and no failover as the primary ones I personally encountered)

The approach I took in this refactor is more of a redis native approach where I tried to remove much of the complexity of the scripts and allow all the election/promotion to be done through the redis-sentinels with a small init script hosted here as a configmap. I feel like this also makes it easier to maintain this chart and allows it to be more dynamic as it can be used with the official redis image or any image really. While I don't think it's perfect, I think this puts the chart in much more of a stable category than it's current state.

I'll update the PR notes to add these details for anyone else looking.

Reply:
@khrisrichardson that's definitely something I was worried about. I tried to have the init script rely on the current state of the cluster to prevent any split/partition issues. So far I haven't noticed through upgrades and what-not but I'm only able to test on a kubespray type of deployment.

Reply:
Hi @khrisrichardson and @ssalaues Thanks for the additional info. I see.

@ssalaues Would you please add an "Upgrading" section to the README summarizing what you said above?

That this chart version changes simplifies strategy, allowing use of the [official image](https://hub.docker.com/_/redis/) which has no such scripts so doesn't need RBAC or SA? And specifically, when upgrading from `>=2.0.1` to `3.0.0`, the `Role`, `RoleBinding`, and `ServiceAccount ` resources associated with the chart should be manually deleted. See https://github.com/helm/charts/issues/5657

Reply:
Ahh that's a really good point @khrisrichardson , so basically "forcing the failover in the case of an existing redis cluster but no master" during init if I understand correctly?

Reply:
Again I'm no expert, but that about sums it up. Just to reiterate the concern: there appears to be a small window between

```bash
MASTER=`redis-cli -h {{ template "redis-ha.fullname" . }} -p {{ .Values.sentinel.port }} \
sentinel get-master-addr-by-name {{ .Values.redis.masterGroupName }} | \
grep -E '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}'`
```

and 

```bash
if [[ ! `redis-cli -h $MASTER ping` ]]; then
   echo "Cannot contact master, replacing dead master with default master"
   setup_defaults
else
    echo "Found reachable master, updating config"
    sentinel_update $MASTER
    redis_update $MASTER
fi
```

during which time sentinel could have initiated a leader election. By having the init container force the failover, it would override what is otherwise happening behind the scenes (supposing that multiple init containers aren't making the same decision concurrently). But at least sentinel is informed of what's happening vs. being surprised by a second node thinking it's master.

Probably better to consult with an expert than leave it to my postulating.

Reply:
Yeah I'm with you on this, I'll test some stuff out. I think I haven't seen any issues in my testing because when upgrading or testing fail overs, the master pod is typically down for longer than the time it takes for Sentinel to fail over (even if you just delete the pod). So I need to test out some scenarios where the master might go down and back up fairly quickly.

Reply:
I'm gonna mark this conversation as resolved since we kinda got off the original subject. But thanks to input from @khrisrichardson I was able to find some corner cases that needed to be fixed and was able to greatly simplify the init script.